zipline/lib/adjustment.pyx

```diff
@@ -111,6 +117,56 @@ cpdef make_adjustment_from_indices(Py_ssize_t first_row,
     return type_(first_row, last_row, first_column, last_column, value)
 
 
+cdef type _choose_adjustment_type(AdjustmentKind adjustment_kind,
+                                  column_type value):
+    if adjustment_kind in (ADD, MULTIPLY):
+        if column_type is np.float64_t:
+            return _float_adjustment_types[adjustment_kind]
+
+        raise TypeError(
+            "Can't construct %s Adjustment with value of type %r.\n"
+            "ADD and MULTIPLY adjustments are only supported for "
+            "floating point data." % (
+                ADJUSTMENT_KIND_NAMES[adjustment_kind],
+                type(value),
+            )
+        )
+
+    elif adjustment_kind == OVERWRITE:
```

Comment:
one is compile time dispatching, the other is runtime dispatching. Cython doesn't really let us merge these paths.